framework/src/tinyalsa/tinyalsa.c

```diff
@@ -1292,82 +1365,95 @@ int pcm_wait(struct pcm *pcm, int timeout)
 	unsigned int size;
 	int prio;
 	struct timespec st_time;
+	int count = 0;
 
 	if (pcm == NULL) {
 		return -EINVAL;
 	}
-	if ((pcm->flags & PCM_OUT) && pcm->buf_ptr < pcm->buffer_cnt) {
-		/* In playback scenario, if audio buffers are available, return immediately */
-		pcm->next_buf = pcm->pBuffers[pcm->buf_ptr];
-		pcm->next_size = pcm->pBuffers[pcm->buf_ptr]->nmaxbytes;
-		return 1;
-	} else {
-		/* We dont have any empty buffers. wait for deque message from kernel */
-		if (timeout > 0) {
-			clock_gettime(CLOCK_REALTIME, &st_time);
-			st_time.tv_nsec += timeout * MILLI_TO_NANO;
-			size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
-		} else {
-			size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
-		}
 
+	/* If there are some messages in the queue, empty them */
+	while (1) {
+		clock_gettime(CLOCK_REALTIME, &st_time);
+		size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
 		if (size != sizeof(msg)) {
-			if (errno == ETIMEDOUT) {
-				oops(pcm, errno, "TIMEOUT while watiting for deque message from kernel");
-				return 0;
-			} else {
-				return oops(pcm, errno, "Interrupted while waiting for deque message from kernel");
-			}
+			break;
 		}
+
+		/* If you find a dequeue message in the queue,
+		update the apb flags and curbyte and increment the count*/
 		if (msg.msgId == AUDIO_MSG_DEQUEUE) {
 			apb = (struct ap_buffer_s *)msg.u.pPtr;
-			pcm->next_buf = apb;
-			pcm->next_size = apb->nbytes;
-			pcm->buf_ptr--;
-			return 1;
+			apb->flags &= ~AUDIO_APB_OUTPUT_ENQUEUED;
+			apb->curbyte = 0;
+			count++;
+		}
+	}
+
+	if (count > 0) {
+		return 1;
+	}
+
+	/* If there were no buffers in the queue, wait for codec to put a buffer on the queue */
+	if (timeout > 0) {
+		/* Use the timeout given by application */
+		clock_gettime(CLOCK_REALTIME, &st_time);
+		st_time.tv_nsec += timeout * MILLI_TO_NANO;
+		size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
+	} else {
+		/* Application did not give a timeout value
+		The behavior here depends on whether mq was opened with O_NONBLOCK
+		This might block until a msg becomes available on the queue */
+		size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
```

Comment:
In my opinion, we should check for pcm_avail_update, because if pcm_wait is called when no buffer is enqueued. (pcm_open(PCM_OUT | PCM_MMAP) -> pcm_wait(pcm,-1)) , the framework would wait indefinitely .

Reply:
True. Fixed.